tree/ntupleutil/v7/src/RNTupleExporter.cxx

```diff
@@ -0,0 +1,167 @@
+   }
+}
+
+int CountPages(const RNTupleDescriptor &desc, std::span<const RColumnExportInfo> columns)
+{
+   int nPages = 0;
+   DescriptorId_t clusterId = desc.FindClusterId(0, 0);
+   while (clusterId != kInvalidDescriptorId) {
+      const auto &clusterDesc = desc.GetClusterDescriptor(clusterId);
+      for (const auto &colInfo : columns) {
+         const auto &pages = clusterDesc.GetPageRange(colInfo.fColDesc->GetPhysicalId());
+         nPages += pages.fPageInfos.size();
+      }
+      clusterId = desc.FindNextClusterId(clusterId);
+   }
+   return nPages;
+}
+
+} // namespace
+
+RExportPagesResult RNTupleExporter::ExportPages(RPageSource &source, const RExportPagesOptions &options)
+{
+   RExportPagesResult res = {};
+
+   // make sure the source is attached
+   source.Attach();
+
+   auto desc = source.GetSharedDescriptorGuard();
+   RClusterPool clusterPool{source};
+
+   // Collect column info
+   std::vector<RColumnExportInfo> columnInfos;
+   AddColumnsFromField(columnInfos, desc.GetRef(), desc->GetFieldZero());
+
+   // Collect ColumnSet for the cluster pool query
+   RCluster::ColumnSet_t columnSet;
+   columnSet.reserve(columnInfos.size());
+   for (const auto &colInfo : columnInfos) {
+      columnSet.emplace(colInfo.fColDesc->GetPhysicalId());
+   }
+
+   const auto nPages = CountPages(desc.GetRef(), columnInfos);
+
+   const bool showProgress = (options.fFlags & RExportPagesOptions::kShowProgressBar) != 0;
+   res.fExportedFileNames.reserve(nPages);
+
+   // Iterate over the clusters in order and dump pages
+   DescriptorId_t clusterId = desc->FindClusterId(0, 0);
+   int pagesExported = 0;
+   int prevIntPercent = 0;
+   while (clusterId != kInvalidDescriptorId) {
+      const auto &clusterDesc = desc->GetClusterDescriptor(clusterId);
+      const RCluster *cluster = clusterPool.GetCluster(clusterId, columnSet);
+      for (const auto &colInfo : columnInfos) {
+         DescriptorId_t columnId = colInfo.fColDesc->GetPhysicalId();
+         const auto &pages = clusterDesc.GetPageRange(columnId);
+         const auto &colRange = clusterDesc.GetColumnRange(columnId);
+         std::uint64_t pageIdx = 0;
+
+         R__LOG_DEBUG(0, RNTupleExporterLog())
+            << "exporting column \"" << colInfo.fQualifiedName << "\" (" << pages.fPageInfos.size() << " pages)";
+
+         // We should never try to export a suppressed column range
+         assert(!colRange.fIsSuppressed || pages.fPageInfos.empty());
+
+         for (const auto &pageInfo : pages.fPageInfos) {
+            ROnDiskPage::Key key{columnId, pageIdx};
+            const ROnDiskPage *onDiskPage = cluster->GetOnDiskPage(key);
+
+            // dump the page
+            const void *pageBuf = onDiskPage->GetAddress();
+            const auto maybeChecksumSize = (options.fFlags & RExportPagesOptions::kIncludeChecksums) ? 8 : 0;
+            const std::uint64_t pageBufSize = pageInfo.fLocator.fBytesOnStorage + maybeChecksumSize;
```

Comment:
Note that we also have to check if the page actually has a checksum (`pageInfo.fHasChecksum`). NB: perhaps we should amend the documentation of `ROnDiskPage` and clarify that this store the page data and the checksum (if applicable).